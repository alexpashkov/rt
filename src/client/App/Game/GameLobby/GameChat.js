import React from 'react';
import PropTypes from 'prop-types';

const GameChat = ({
  messages = [],
  message,
  handleMessageChange,
  handleMessageSend
}) => (
  <div>
    <main>
      {messages.length
        ? messages.map((message, i) => <pre key={i}>{JSON.stringify(message)}</pre>)
        : 'Be the first who will send a message!'}
    </main>
    <form onSubmit={handleMessageSend}>
      <input
        type="text"
        value={message}
        onChange={handleMessageChange}
      />
      <button>Send</button>
    </form>
  </div>
);

GameChat.propTypes = {
  messages: PropTypes.array.isRequired,
  message: PropTypes.string.isRequired,
  handleMessageChange: PropTypes.func.isRequired,
  handleMessageSend: PropTypes.func.isRequired
};

export default GameChat;
