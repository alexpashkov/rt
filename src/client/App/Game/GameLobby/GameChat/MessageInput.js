import React from 'react';
import PropTypes from 'prop-types';

const MessageInput = ({
  currentMessage,
  setCurrentMessage,
  handleMessageSending
}) => (
  <form onSubmit={handleMessageSending}>
    <input
      type="text"
      value={currentMessage}
      onChange={event => setCurrentMessage(event.currentTarget.value)}
    />
    <button>Send</button>
  </form>
);

MessageInput.propTypes = {
  currentMessage: PropTypes.string.isRequired,
  setCurrentMessage: PropTypes.func.isRequired,
  handleMessageSending: PropTypes.func.isRequired,
};

export default MessageInput;
