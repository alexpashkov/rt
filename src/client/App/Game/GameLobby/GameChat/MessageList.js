import React from 'react';
import PropTypes from 'prop-types';

const MessageList = ({ messages }) => (
  <div>
    {messages.length
      ? messages.map((message, i) => (
          <pre key={i}>{JSON.stringify(message)}</pre>
        ))
      : 'Be the first who will send a message!'}
  </div>
);

MessageList.propTypes = {
  messages: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      login: PropTypes.string.isRequired,
      message: PropTypes.string.isRequired
    })
  )
};

export default MessageList;
