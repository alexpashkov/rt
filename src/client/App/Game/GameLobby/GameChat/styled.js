import styled from 'react-emotion';
import { lighten } from 'polished';

export const GameChatWrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

export const MessagesWrapper = styled.div`
  flex-grow: 1;
  overflow: auto;
  height: 30vh;
  &:empty {
    display: flex;
    justify-content: center;
    align-items: center;
    &::after {
      content: 'Be the first in the chat!';
      color: ${props => props.theme.colors.secondaryText};
      display: block;
    }
  }
`;

export const MessageItemWrapper = styled.div`
  padding: 5px;
  display: flex;
`;
export const UserIdWrapper = styled.div`
  align-self: flex-end;
  margin-top: 20px;
  height: 40px;
  width: 40px;
  background-color: ${props => props.theme.colors.accent};
  border-radius: 10px;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 10px;
`;
export const MessageWrapper = styled.div`
  align-self: flex-start;
  flex-grow: 1;
  margin-bottom: 20px;
  margin-left: 10px;
  background-color: ${props => lighten(0.1, props.theme.colors.background)};
  padding: 15px;
  border-radius: 10px;
`;
