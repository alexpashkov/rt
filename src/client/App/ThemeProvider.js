import React from 'react';
import { ThemeProvider as EmotionThemeProvider } from 'emotion-theming';
import theme from '../theme';

export const ThemeProvider = ({ children }) => (
  <EmotionThemeProvider theme={theme}>{children}</EmotionThemeProvider>
);

export default ThemeProvider;
