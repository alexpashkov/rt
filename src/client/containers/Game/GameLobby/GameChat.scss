@import "../../../styles/variables";

.GameChat {
  border: 1px solid darken($mainBGColor, 10%);
  padding: 15px;
  border-radius: 5px;
}

.GameChat__messages {
  height: 300px;
  overflow-y: auto;
}

.GameChat__message {
  margin: 10px;
  display: flex;
  align-items: flex-end;
}

.GameChat__message-sender {
  width: 60px;
  overflow: hidden;
  text-overflow: ellipsis;
}

.GameChat__message-content {
  margin-left: 10px;
  width: 100%;
  background-color: darken($mainBGColor, 10%);
  padding: 15px;
}


.GameChat__message-input-form {
  display: flex;
}

.GameChat__message-input {
  flex-grow: 1;
  background: none;
  border: none;
  border-bottom: 1px solid darken($mainBGColor, 10%);
  line-height: 1.15;
}

.GameChat__message-send-button {
  margin-left: 10px;
}
