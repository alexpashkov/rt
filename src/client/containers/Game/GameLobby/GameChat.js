import React, { Component } from "react";
import { compose } from "ramda";
import withSocket from "../../../hocs/with-socket";
import { connect } from "react-redux";
import { appendChatMessage } from "../../../actions/game-chat-messages";
import {
  client as clientSocketEvents,
  server as serverSocketEvents
} from "../../../../shared/types";
import "./GameChat.scss";

class GameChat extends Component {
  state = {
    message: ""
  };

  componentDidUpdate({messages: prevMessages = []}) {
    const { messages = [] } = this.props;
    if (prevMessages.length !== messages.length) {
      this.messagesElem.scrollTop = this.messagesElem.scrollHeight;
    }
  }

  render() {
    const { messages } = this.props;
    return (
      <section className="GameChat">
        <h2>Game Chat</h2>
        <div ref={el => (this.messagesElem = el)} className="GameChat__messages">
          {messages.map(({ id, login, message }) => (
            <article key={id} className="GameChat__message">
              <div className="GameChat__message-sender">{login}</div>
              <div className="GameChat__message-content">{message}</div>
            </article>
          ))}
        </div>
        <form className="GameChat__message-input-form" onSubmit={this.handleMessageSend}>
          <input
            type="text"
            className="GameChat__message-input"
            onChange={this.handleMessageChange}
            value={this.state.message}
          />
          <button className="GameChat__message-send-button primary">Send</button>
        </form>
      </section>
    );
  }
  handleMessageChange = ({ target: { value } }) => {
    this.setState({
      message: value
    });
  };
  handleMessageSend = e => {
    e.preventDefault();
    const { socket } = this.props;
    socket.emit(clientSocketEvents.GAME_CHAT_MESSAGE, this.state.message);
    this.setState({message: ""})
  };
  componentDidMount() {
    const { socket, appendChatMessage } = this.props;
    socket.on(serverSocketEvents.GAME_CHAT_MESSAGE, appendChatMessage);
  }
  componentWillUnmount() {
    const { socket, appendChatMessage } = this.props;
    socket.removeListener(
      serverSocketEvents.GAME_CHAT_MESSAGE,
      appendChatMessage
    );
  }
}

export default compose(
  withSocket,
  connect(
    state => ({
      messages: state.game.messages
    }),
    {
      appendChatMessage
    }
  )
)(GameChat);
