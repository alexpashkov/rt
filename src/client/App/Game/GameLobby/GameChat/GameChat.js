import React from 'react';
import PropTypes from 'prop-types';

import MessageList from './MessageList';
import MessageInput from './MessageInput';
import { GameChatWrapper } from './styled';

const GameChat = ({
  messages = [],
  currentMessage,
  setCurrentMessage,
  handleMessageSending,
  className
}) => (
  <GameChatWrapper className={className}>
    <MessageList messages={messages} />
    <MessageInput
      currentMessage={currentMessage}
      setCurrentMessage={setCurrentMessage}
      handleMessageSending={handleMessageSending}
    />
  </GameChatWrapper>
);

GameChat.propTypes = {
  messages: PropTypes.array.isRequired,
  currentMessage: PropTypes.string.isRequired,
  setCurrentMessage: PropTypes.func.isRequired,
  handleMessageSending: PropTypes.func.isRequired,
  className: PropTypes.string
};

export default GameChat;
