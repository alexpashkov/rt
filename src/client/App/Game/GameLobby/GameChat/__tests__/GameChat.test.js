import React from "react";
import GameChat from '../GameChat';
import renderer from 'react-test-renderer';
import ThemeProvider from "../../../../ThemeProvider"

describe('GameChat', () => {
  it('Renders without throwing an exception', () => {
    const component = renderer.create(<ThemeProvider><GameChat /></ThemeProvider>);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
