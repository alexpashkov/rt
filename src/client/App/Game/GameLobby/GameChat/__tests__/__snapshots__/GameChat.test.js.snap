// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameChat Renders without throwing an exception 1`] = `
<div
  className="css-j7qwjs euynx710"
>
  <div
    className="css-cre6xo euynx711"
  />
  <form
    className="css-1q4kx8i ejf5oyo0"
    onSubmit={undefined}
  >
    <input
      className="css-1s87wor ejf5oyo1"
      onChange={[Function]}
      type="text"
      value={undefined}
    />
    <button
      className="css-13km75h e137ia0s0"
      color="primary"
      size="sm"
      type="submit"
    >
      Send
    </button>
  </form>
</div>
`;
